Add Go function to create serialized ConfigOptions protos

frreiss opened the pull request:
As noted in #22926, the only way to specify session options when creating a session from the Go API involves running a Python program to generate the serialized byte representation of a ConfigOptions protobuf, then pasting those bytes into a `[]byte` literal in the Go code.

This PR adds a Go function, `NewConfigOptions`, that generates a serialized `ConfigOptions` protocol buffer message based on parameters provided by the caller. To minimize impact on the rest of the TensorFlow code, the implementation of `NewConfigOptions` relies on the experimental C API call `TF_CreateConfig`. As a result, only the options exposed in `TF_CreateConfig` are available. 

Users can use `NewConfigOptions` to initialize the `Config` field of a `SessionOptions` struct, then pass the struct to `NewSession()`:
```go
config := NewConfigOptions(true, true, 1)
opts := SessionOptions{Config: config}
s, err := NewSession(graph, &opts)
```

I did not change the any existing Go code that passes ConfigOptions as serialized byte strings directly will still work; for example:
```go
opts := SessionOptions{Config: []byte("(\x01")}
s, err := NewSession(graph, &opts)
```

frreiss committed:
Add function to create serialized ConfigOptions protos

asimshankar reviewed (commented):
Thanks for doing this!

I think it might be better to define a `struct` instead of a `func`, which will make it easier to add/remove/update options in the future and also make it easier to read the code (`true, true, 0` isn't that easy to follow ;)). Something like:

```go
type Config struct {
  XLACompile bool
  AllowGPUMemoryGrowth bool
}

func (c *Config) Bytes() []byte {
  // Mostly what you have in NewConfigOptions right now
}
```

Which would cause the test to look like:

```go
config := SessionConfig{XLACompile: true, GPUAllowMemoryGrowth: true, NumCPUs: 1}
opts := SessionOptions{Config: config.Bytes()}
```
instead of:
```go
config := NewConfigOptions(true, true, 1)
```

Thoughts?

frreiss commented:
> Thanks for doing this!
> 
> I think it might be better to define a `struct` instead of a `func`, which will make it easier to add/remove/update options in the future and also make it easier to read the code (`true, true, 0` isn't that easy to follow ;)). Something like:
> 
> ```go
> type Config struct {
>   XLACompile bool
>   AllowGPUMemoryGrowth bool
> }
> 
> func (c *Config) Bytes() []byte {
>   // Mostly what you have in NewConfigOptions right now
> }
> ```
> 
> Which would cause the test to look like:
> 
> ```go
> config := SessionConfig{XLACompile: true, GPUAllowMemoryGrowth: true, NumCPUs: 1}
> opts := SessionOptions{Config: config.Bytes()}
> ```
> 
> instead of:
> 
> ```go
> config := NewConfigOptions(true, true, 1)
> ```
> 
> Thoughts?

Great idea! Let me update this PR to use a struct. It'll make it easier to adjust to changes in `TF_CreateConfig`. Also I can add some sensible default values too.

frreiss committed:
Introduce Config struct to encapsulate session config options

frreiss commented:
Update: I've added a new struct `Config` to encapsulate the session configuration parameters. I've also added logic to ensure that the zero values for all the fields in the `Config` struct correspond to the TensorFlow defaults of the associated parameters.
I also removed some outdated comments from `session_test.go` that I should have removed earlier.

asimshankar commented on tensorflow/go/session.go:
```diff
@@ -349,6 +350,79 @@ func (o *SessionOptions) c() (ret *C.TF_SessionOptions, done func(), err error)
+	// DEFAULT is the default setting for this version of TensorFlow,
+	// Currently the default is OFF, but it will change to ON in a future
+	// version.
+	DEFAULT JitLevel = 0
```
Minor nit: Let's use mixed case naming for these as is more idiomatic in go (https://golang.org/doc/effective_go.html#mixed-caps and constants in standard library package like https://golang.org/pkg/net/http/#pkg-constants). Also "Default", "On", and "Off" are pretty generic names, how about:

```go
const (
  JitDefault JitLevel = 0
  JitOn JitLevel = 1
  JitOff JitLevel = -1
)
```

frreiss replied:
Sure, no problem. Making those changes now.

asimshankar reviewed (dismissed):
Thanks, looks good. One minor naming suggestion, otherwise looks good.

@jhseu @alextp would need someone from the TensorFlow team to merge it in. Thanks!

alextp commented:
Please address @asimshankar's comments and I'll approve / merge this.

frreiss committed:
Address review comments and add some more API docs

frreiss commented:
I've pushed an additional commit that should address the review comments. I also updated the API docs for `SessionOptions.Config`.

Please let me know if there is anything else I need to do on this PR.

BIGCRYPTO001 commented:
Hello

mattn commented:
Why this is reverted?

https://github.com/tensorflow/tensorflow/commit/6e9cb400d17f60e50cefe59fc099fc5b6e4074ce#diff-dcd28ad951bd17e9c512d3b564640bab

bedbugs-bit commented:
If the the branch has been merged and the problem solved, why the issue #22926  still open?

Rahulraj0308 commented:
directly configure TensorFlow session options in Go without relying on serialized protobuf
package main

import (
	"fmt"
	"log"

	tf "github.com/tensorflow/tensorflow/tensorflow/go"
)

func main() {
	// Define your session options directly
	sessionOptions := &tf.SessionOptions{
		Target:       "local", // Set your target device, e.g., "local" or "gpu:0"
		Config: &tf.ConfigProto{
			AllowSoftPlacement: true, // Set your configuration options
			LogDevicePlacement: true,
			InterOpParallelismThreads: 1,
			IntraOpParallelismThreads: 1,
		},
	}

	// Load your graph (replace with your graph loading code)
	graph := loadGraph()

	// Create a new session with the specified options
	session, err := tf.NewSession(graph, sessionOptions)
	if err != nil {
		log.Fatal(err)
	}
	defer session.Close()

	// Your TensorFlow operations go here...

	fmt.Println("TensorFlow session successfully created!")
}

func loadGraph() *tf.Graph {
	// Replace this function with your graph loading code
	// Example: Load a graph from a SavedModel
	model, err := tf.LoadSavedModel("/path/to/saved_model", []string{"serve"}, nil)
	if err != nil {
		log.Fatal(err)
	}

	return model.Graph
}
you directly configure the SessionOptions struct with the desired settings for your TensorFlow session.

Rahulraj0308 commented:
One more thing you can do is to directly configure TensorFlow session options in Go without relying on serialized `protobuf....`
`package main

import (
	"fmt"
	"log"

	tf "github.com/tensorflow/tensorflow/tensorflow/go"
)

func main() {
	
	sessionOptions := &tf.SessionOptions{
		Target:       "local",
		Config: &tf.ConfigProto{
			AllowSoftPlacement: true, 
			LogDevicePlacement: true,
			InterOpParallelismThreads: 1,
			IntraOpParallelismThreads: 1,
		},
	}

	
	graph := loadGraph()

	
	session, err := tf.NewSession(graph, sessionOptions)
	if err != nil {
		log.Fatal(err)
	}
	defer session.Close()

	

	fmt.Println("TensorFlow session successfully created!")
}

func loadGraph() *tf.Graph {
	
	model, err := tf.LoadSavedModel("/path/to/saved_model", []string{"serve"}, nil)
	if err != nil {
		log.Fatal(err)
	}

	return model.Graph
}
`
you directly configure the SessionOptions struct with the desired settings for your TensorFlow session.
